substrate/client/network/src/litep2p/shim/notification/tests/peerset.rs

```diff
@@ -0,0 +1,812 @@
+
+			for peer in &out_peers {
+				assert_eq!(
+					peerset.peers().get(&peer),
+					Some(&PeerState::Opening { direction: Direction::Outbound(Reserved::No) })
+				);
+			}
+
+			out_peers[0]
+		},
+		event => panic!("invalid event: {event:?}"),
+	};
+
+	// disconnect the now-opening peer
+	to_peerset.unbounded_send(PeersetCommand::DisconnectPeer { peer }).unwrap();
+
+	// poll `Peerset` to register the command and verify the peer is now in state `Canceled`
+	futures::future::poll_fn(|cx| match peerset.poll_next_unpin(cx) {
+		Poll::Pending => Poll::Ready(()),
+		_ => panic!("unexpected event"),
+	})
+	.await;
+
+	assert_eq!(
+		peerset.peers().get(&peer),
+		Some(&PeerState::Canceled { direction: Direction::Outbound(Reserved::No) })
+	);
+
+	// the substream failed to open, verify that peer state is now `Backoff`
+	// and that `Peerset` doesn't emit any events
+	peerset.report_substream_open_failure(peer, NotificationError::NoConnection);
+	assert_eq!(peerset.peers().get(&peer), Some(&PeerState::Backoff));
+
+	futures::future::poll_fn(|cx| match peerset.poll_next_unpin(cx) {
+		Poll::Pending => Poll::Ready(()),
+		_ => panic!("unexpected event"),
+	})
+	.await;
+}
+
+#[tokio::test]
+async fn peer_disconnected_when_being_validated_then_accepted() {
+	sp_tracing::try_init_simple();
+
+	let peerstore_handle = Arc::new(peerstore_handle_test());
+	let _reserved = (0..1)
+		.map(|_| {
+			let peer = PeerId::random();
+			peerstore_handle.add_known_peer(peer);
+			peer
+		})
+		.collect::<Vec<_>>();
+	let (mut peerset, to_peerset) = Peerset::new(
+		ProtocolName::from("/notif/1"),
+		25,
+		25,
+		false,
+		Default::default(),
+		Default::default(),
+		peerstore_handle,
+	);
+	assert_eq!(peerset.num_in(), 0usize);
+	assert_eq!(peerset.num_out(), 0usize);
+
+	let peer = match peerset.next().await {
+		Some(PeersetNotificationCommand::OpenSubstream { peers: out_peers }) => {
+			assert_eq!(peerset.num_in(), 0usize);
+			assert_eq!(peerset.num_out(), 1usize);
+			assert_eq!(out_peers.len(), 1);
+
+			for peer in &out_peers {
+				assert_eq!(
+					peerset.peers().get(&peer),
+					Some(&PeerState::Opening { direction: Direction::Outbound(Reserved::No) })
+				);
+			}
+
+			out_peers[0]
+		},
+		event => panic!("invalid event: {event:?}"),
+	};
+
+	// disconnect the now-opening peer
+	to_peerset.unbounded_send(PeersetCommand::DisconnectPeer { peer }).unwrap();
+
+	// poll `Peerset` to register the command and verify the peer is now in state `Canceled`
+	futures::future::poll_fn(|cx| match peerset.poll_next_unpin(cx) {
+		Poll::Pending => Poll::Ready(()),
+		_ => panic!("unexpected event"),
+	})
+	.await;
+
+	assert_eq!(
+		peerset.peers().get(&peer),
+		Some(&PeerState::Canceled { direction: Direction::Outbound(Reserved::No) })
+	);
+
+	// the substream failed to open, verify that peer state is now `Backoff`
+	// and that `Peerset` doesn't emit any events
+	peerset.report_substream_open_failure(peer, NotificationError::NoConnection);
```

Comment:
As the name of the test suggest, probably this should be `peerset.report_substream_opened()`.

Reply:
Fixed